CMakeLists update for issue 148

ndellingwood opened the pull request:
Address request in issue #148 to check for
CMAKE_CXX_USE_RESPONSE_FILE_FOR_OBJECTS=ON when Cuda is enabled. 
If it is not on, error out at configuration.

ndellingwood commented:
Tested that this 1) errors out at configuration with Cuda8 if `CMAKE_CXX_USE_RESPONSE_FILE_FOR_OBJECTS` is not set on and 2) that it configures properly when the variable is set.

mhoemmen commented on CMakeLists.txt:
```diff
@@ -75,6 +75,10 @@ TRIBITS_ADD_OPTION_AND_DEFINE(
     LIST(APPEND DEVICE_LIST "<Cuda,CudaUVMSpace>")
   ENDIF()
 
+  IF(NOT CMAKE_CXX_USE_RESPONSE_FILE_FOR_OBJECTS)
```
```
IF(NOT (DEFINED(CMAKE_CXX_USE_RESPONSE_FILE_FOR_OBJECTS)) OR (NOT CMAKE_CXX_USE_RESPONSE_FILE_FOR_OBJECTS))
```
etc.

mhoemmen replied:
Not sure if `CMAKE_CXX_USE_RESPONSE_FILE_FOR_OBJECTS` is defined by default.  (CMake options have three possible states: `NOT DEFINED`, `FALSE`, and `TRUE`.)  The above will check for that.

ndellingwood replied:
@mhoemmen I didn't realize that, thanks for the catch!

srajama1 commented:
Is there a problem with always having this ON when CUDA is enabled ? I don't know why we would error out, force the user to set it and reconfigure again. If it is needed can we just set it ON for CUDA.

mhoemmen commented on CMakeLists.txt:
```diff
@@ -75,6 +75,10 @@ TRIBITS_ADD_OPTION_AND_DEFINE(
   ENDIF()
 
+  IF(NOT CMAKE_CXX_USE_RESPONSE_FILE_FOR_OBJECTS)
+    MESSAGE( FATAL_ERROR "Kokkos-Kernels configuration failed: CMAKE_CXX_USE_RESPONSE_FILE_FOR_OBJECTS is required be set to ON when compiling with Cuda." )
```
Incorrect grammar; try this:
```
The CMake option CMAKE_CXX_USE_RESPONSE_FILE_FOR_OBJECTS is either undefined or OFF.  Please set CMAKE_CXX_USE_RESPONSE_FILE_FOR_OBJECTS:BOOL=ON when building with CUDA.
```

ndellingwood replied:
@mhoemmen made the change, thanks!

ndellingwood commented:
Updated PR with suggested changes

ndellingwood commented:
@bartlettroscoe I added a check in kokkos-kernels for `CMAKE_CXX_USE_RESPONSE_FILE_FOR_OBJECTS` when Cuda is enabled; should this check be moved within Trilinos and, if Cuda and kokkos-kernels are enabled, turn `CMAKE_CXX_USE_RESPONSE_FILE_FOR_OBJECTS=ON` by default so users don't have to reconfigure like @srajama1 mentioned?

bartlettroscoe commented:
> @bartlettroscoe I added a check in kokkos-kernels for CMAKE_CXX_USE_RESPONSE_FILE_FOR_OBJECTS when Cuda is enabled; should this check be moved within Trilinos and, if Cuda and kokkos-kernels are enabled, turn CMAKE_CXX_USE_RESPONSE_FILE_FOR_OBJECTS=ON by default so users don't have to reconfigure like @srajama1 mentioned?

This file gets processed as part of the Trilinos CMake configure as well so it should be active there as well.  In fact, any TriBITS configure of this package will have this check.

mhoemmen commented:
Thanks @ndellingwood ! :D

ndellingwood commented:
Thanks @mhoemmen!
1. I don't believe the Cuda build with complex OFF requires this, I can update so the message and check is more specific. 
2. I have not tested against all supported host compilers, though from the Tribits documentation the possible issues involving use of `CMAKE_CXX_USE_RESPONSE_FILE_FOR_OBJECTS` seem to stem from combination of OS + CMake version + compiler. I'm not sure how to adequately address this, as testing of host compilers from sems modules on a Linux OS may give false confidence that a compiler is compatible, whereas another combo of CMake + OS may fail.

mhoemmen commented:
@ndellingwood wrote:
> I don't believe the Cuda build with complex OFF requires this, I can update so the message and check is more specific.

I'm not too worried about this bit, as long as the build settings work in both complex ON and OFF cases, but it's still helpful to know, so thanks :-) .

> I'm not sure how to adequately address this, as testing of host compilers from sems modules on a Linux OS may give false confidence that a compiler is compatible, whereas another combo of CMake + OS may fail.

If we don't test it, it doesn't exist ;-) .  I would say, let's discover this rather than trying to guess.  If somebody reports on a new platform that it doesn't work, we can change the message then.

ndellingwood commented:
@crtrott requested the `FATAL_ERROR` be changed to `WARNING` in `CMakeLists.txt`
Updated the PR with improved message and more restricted check (Cuda and complex double enabled) as suggested, and changed replaced with `WARNING`. 
Since the `FATAL_ERROR` was replaced by a `WARNING`  no one will be forced to use this option which potentially may not work with some combo of host compiler + OS + CMake version. 

@srajama1 since `CMAKE_CXX_USE_RESPONSE_FILE_FOR_OBJECTS` will not be enabled by default (discussion and links in issue #176), I think this PR is ready.

ndellingwood commented:
@srajama1 README updated with comment for usage `CMAKE_CXX_USE_RESPONSE_FILE_FOR_OBJECTS` when building with Trilinos.

mhoemmen commented on CMakeLists.txt:
```diff
@@ -75,6 +75,10 @@ TRIBITS_ADD_OPTION_AND_DEFINE(
     LIST(APPEND DEVICE_LIST "<Cuda,CudaUVMSpace>")
   ENDIF()
 
+  IF( Trilinos_ENABLE_COMPLEX_DOUBLE AND ((NOT DEFINED CMAKE_CXX_USE_RESPONSE_FILE_FOR_OBJECTS) OR (NOT CMAKE_CXX_USE_RESPONSE_FILE_FOR_OBJECTS)) )
```
This is fine, though it may be possible to turn OFF `KokkosKernels_INST_COMPLEX_DOUBLE` even if `Trilinos_ENABLE_COMPLEX_DOUBLE` is ON.

ndellingwood replied:
@mhoemmen Thanks for the catch! I'm guessing most users use `Trilinos_ENABLE_COMPLEX_DOUBLE` so I think I should leave that and add a check for  KokkosKernels_INST_COMPLEX_DOUBLE also, does this sound correct?

mhoemmen replied:
@ndellingwood What about stand-alone builds?  Does `Trilinos_ENABLE_COMPLEX_DOUBLE` even exist in that case?

ndellingwood replied:
@mhoemmen I'm not sure that kokkos-kernels has standalone CMake support, I think for standalone builds the Makefile system should be used but I may be wrong.
